### ISSUE TITLE
Create Typescript definition file for Angular 2 testing

### ISSUE THREAD

**Issue by marcysutton** [dequelabs/axe-core#183]
To support Angular 2 development with axe-core, we should create a Typescript definition file for type annotations and hints. I still have to look into specifics, but I wanted to capture this now. Here is some documentation on creating a type definition file: 
- http://definitelytyped.org/guides/creating.html
- http://peter.grman.at/how-to-write-typescript-definition-files/

**Comment by marcysutton** [comment]
Things I've found so far that we may want to address:

- We specify optional parameters before required ones in `axe.a11yCheck` and `axe._runRules`, which isn't allowed in Typescript. ( i.e. `context: any, options?:Object, callback:Function`).
- There are multiple methods and objects exposed in the global namespace that we may not want to be called directly. (i.e. `_audit`, `_load()`, `_runRules()`, `commons`, `constants`, `utils`, etc.)
- Arguments aren't always consistent, i.e. `callback` vs. `cb` in `axe.a11yCheck` and `axe.addReporter`, respectively.
- We could really use additional code comments and documentation for the newer parts of the API, such as the Plugin system and `axe.configure`.

I'm finding the exercise of creating a Typescript definition file to be super valuable, as it maps our API to interfaces and types. aXe is pretty loose with types in many places, taking full advantage of JavaScript's dynamic nature but making it more challenging to define the API in a definition file. We can work around this with the `any` type, but it seems like a good opportunity to look more closely at our code structure and the public API.

**Comment by dylanb** [comment]
Any method starting with _ is not public for sure

commons and utils are intended to be used in the rules, so they are public but we may need to curate the list to make sure we are not exposing too much.

We should not change existing methods to preserve backwards compatibility. But we already plan to deprecate a11yCheck in favor of a run method that has a better signature.

Can you simply always call the callback the same thing in the TypeScript definition regardless of what it is called in our code? If not, then TypeScript is quite seriously flawed.

**Comment by DanielRosenwasser** [comment]
Hey @marcysutton I saw [this tweet](https://twitter.com/marcysutton/status/755172506089066496). If you need, I'd be happy to help. Is this issue a good place for discussion, or is something like Gitter better?

**Comment by marcysutton** [comment]
Hi @DanielRosenwasser! Either works. This is what I've got so far, I know there are a few kinks to be ironed out with enums and such but I couldn't even get it to work with a simple exported function. https://gist.github.com/marcysutton/a3e84fa229a90258fb6d86171b840d86

**Comment by DanielRosenwasser** [comment]
@marcysutton take a look at my changes here: https://gist.github.com/DanielRosenwasser/1ff72a78965b9790af0cc3426fdc53b7

Some of the changes I made:

First, I switched `axe` to a namespace instead of a `var`. You might've been confused between type and namespace syntax and meant to write `declare var axe: { ... }`, but functions can only be exported from namespaces and objects.

Next, I used `any` instead of `Object`. You almost never want to use `Object` if you can grab any property off of it. Typically, when people write `Object`, they either mean `any` or `{}`. If users of the `.d.ts` might be getting values of some type back, and any properties might be attached to those objects, then `any` is what you want. Otherwise, either is fine.

I also fixed cases where you probably meant to write

* `Plugin` instead of `{Plugin}` and 
* `{ id: string }[]` instead of `[{ id: string }]`

Because those were likely to be errors. The reason is that `{ Plugin }` is the same as writing`{ Plugin: any }`, and `[{ id: string }]` is a single-element tuple containing a `{ id: string }`

Finally I also used array shorthand syntax (i.e. `foo[]`) because it's nicer on the eyes, and normalized between spaces/tabs (and you forced me to use tabs instead of spaces to keep the diff reasonable :smile: ).

Hopefully that helps a little bit!

**Comment by marcysutton** [comment]
This is great, thanks for the info @DanielRosenwasser. Another question as I try to establish a workflow: how do I reference `axe` in a declaration file test? I have a second file called `axe-core-tests.ts`:

```
/// <reference path="axe-core.d.ts" />

axe.a11yCheck() 
```

When I run `ntsc --noImplicitAny typings/axe-core/axe-core-tests.ts` with your latest changes (after a fresh ntypescript install), I get this error:  

```
typings/axe-core/axe-core-tests.ts(4,1): error TS2304: Cannot find name 'axe'.
```

It's probably something obvious...but if you have any tips on testing declaration files, that would be really helpful.

**Comment by DanielRosenwasser** [comment]
@marcysutton Right, I should have mentioned this since Axe seems to introduce a global variable for the browser.

If you want to write your test using the global (and if you want to enable your users to use the `axe` global), then for TypeScript 2.0, you'll add this in place of the `export = axe` line:

```ts
// This is the shape of the module.
export = axe;

// The module is also available as the following global
export as namespace axe;
```

If you're just trying to get it working for TS 1.8, move the `export =` into an ambient module declaration:

```ts
// Outlining the shape of axe globally
declare namespace axe {
    // ...
}

// axe is also available as a module
declare module "axe-core" {
    export = axe;
}
```

If you're planning to publish the package soon, I'd use the latter for now.

**Comment by marcysutton** [comment]
@DanielRosenwasser I was wondering if you could take a look at my PR for the axe-core definition file: https://github.com/dequelabs/axe-core/issues/224 

Am I missing anything? I'm working on consuming it in an Angular 2 app now, which is hard to do without it being published on DefinitelyTyped. Otherwise, tests are passing and it should be ready to submit.
